File: src/pann/Core/Net.cpp
//Net.cpp

#include "Runner.h"
#include "Neuron.h"

#include "Net.h"

using namespace std;

namespace pann
{
    Net::Net()
    {
        setWorkThreadsCount(0);
    } //Net

    Net::~Net()
    {
    } //~Net

    void
    Net::addInputNeuron(NeuronPtr _neuron)
    {
        cache.touch();

        inputNeurons.push_back(_neuron);
    } //addInputNeuron

    void
    Net::removeNeuron(NeuronPtr _neuron)
    {
        cache.touch();

        //Delete all connections to/from current neuron
        BOOST_FOREACH(const Link& link, _neuron->links_in)
            link.getTo()->delOutConnection(_neuron);

        BOOST_FOREACH(const Link& link, _neuron->links_out)
            link.getTo()->delInConnection(_neuron);

        //Remove neuron from registers
        inputNeurons.remove(_neuron);

        Debug()<<_neuron.use_count()<<'\n';
    } //delNeuron

    WeightPtr
    Net::addConnection(NeuronPtr _from, NeuronPtr _to, WeightPtr _weight)
    {
        cache.touch();

        if(!_weight)
            _weight.reset(new Weight());

        _from->addOutConnection(_to, _weight);
        _to->addInConnection(_from, _weight);

        return _weight;
    } //addConnection

    void
    Net::delConnection(NeuronPtr _from, NeuronPtr _to)
    {
        cache.touch();

        /* Short unsignedroduction:
         * Neuron_from               Neuron_to
         *  Link                      Link
         *    to =>Neuron2              to =>Neuron1
         *    w  =>weight1              w  =>weight1
         *    dir=>out                  dir=>in
         */

        _from->delOutConnection(_to);
        _to->delInConnection(_from);
    } //delConnection

    void
    Net::setInput(const valarray<Float>& _input)
    {
        if(_input.size() < inputNeurons.size())
            throw SizeMismatch()<<"setInput(): Supplied input size is smaller "
                                     "then number of input neurons\n";

        if(_input.size() > inputNeurons.size())
            Warning()<<"setInput(): Input size is bigger then input neurons count. "
                             "Check getInputMap() output\n";

        unsigned i = 0;
        BOOST_FOREACH( const NeuronPtr& n, inputNeurons)
           n->setInput(_input[i++]);
    } //setInput

    void
    Net::getOutput(valarray<Float>& _output) const
    {
        if(!cache.isOk())
            regenerateCache();

        unsigned output_size = cache.layers.back().size();
        _output.resize(output_size);

        for(unsigned i = 0; i < output_size; ++i)
            _output[i] = cache.layers.back()[i]->getOutput();
    } //getOutput

    void
    Net::run(Runner& _runner)
    {
        if( !cache.isOk() )
            regenerateCache();

        boost::thread_group threadPool;
        boost::barrier barrier(workThreads);

        //We must give parameters by pointer, because boost will copy all arguments to thread
        for(unsigned thread = 0; thread < workThreads; ++thread)
            threadPool.add_thread( new boost::thread(Net::threadBase, &_runner, this, thread, &barrier) );

        //wait for threads to finish
        threadPool.join_all();
    } //run

    const NetCache&
    Net::getCache() const
    {
        if(!cache.isOk())
            regenerateCache();

        return cache;
    } //getCache

    unsigned 
    Net::getWorkThreadsCount() const
    {
        return workThreads;
    } //getWorkThreadsCount

    void Net::setWorkThreadsCount(unsigned _count)
    {
        if(_count <= 0)
            workThreads = boost::thread::hardware_concurrency();
        else
            workThreads = _count;

    } //setWorkThreadsCount

    void
    Net::formatFront(list<NeuronPtr>& _raw) const
    {
        _raw.sort();
        _raw.unique();

        if(_raw.size() > 0)
        {
            cache.layers.push_back(vector<NeuronPtr>());
            for(list<NeuronPtr>::iterator it = _raw.begin(); it != _raw.end(); ++it)
                cache.layers.back().push_back(*it);
        }
    } //formatFront

    void
    Net::regenerateCache() const
    {
        //TODO: if user set's latency > 1 for regular links - shit may happen
        //TODO: Latency - is very dangerous tool. It's not difficult to understand, 
        //TODO: why this feature is still not implemented in Neuron/Net interface :)
        cache.flush();

        //Here we will place neuron's IDs that will become front, with duplicates
        list<NeuronPtr> rawFront;

        /*
         * Function operates with "hops" attribute of every Neuron
         * to build adequate cache.
         */
        map<NeuronPtr, unsigned> hops;

        //Put inputNeurons to front
        BOOST_FOREACH( const NeuronPtr& n, inputNeurons )
        {
            rawFront.push_back(n);
            hops[n] = 1;
        }
        formatFront(rawFront);

        /*
         * Cache looks like this:
         * At this point cache.layers[0] is the only record and it contains input neurons
         */
        do {
            /*
             * We have to prepare next front
             * At first iteration, if you remember, vector<NeuronPtr> rawFront contains unique inputNeurons
             */
            unsigned nCount = rawFront.size();
            for(unsigned i = 0; i < nCount; ++i)
            {
                //pop_front emulation
                NeuronPtr currentNeuron = rawFront.front();
                rawFront.pop_front();

                //ok, we've got cur_neuron. We will iterate through his Out links
                //and push_back their opposite sides to rawFront
                BOOST_FOREACH( const Link& link, currentNeuron->links_out )
                {
                    const NeuronPtr& toNeuron = link.getTo(); 
                    /*
                     * Short comment. Consider following topology (hops are in brackets):
                     *   +(0)    <-- output          +(0)             +(3)    <-- front(3)
                     *  / \                         / \              / \
                     * +(0)+(0)         ======>    +(0)+(0)  ====>  +(2)+(2)  <-- front(2)
                     *  \ /           iteration1    \ /        N     \ /
                     *   +(0)  <-- input             +(1)             +(1)    <-- front on iteration1
                     *
                     *   For current neuron (C) and other neuron (T), hops might be:
                     *   T = 0  - T-neuron is fresh. We will set T=C+1
                     *   T = N, N < C - it is recurrent link. Don't touch T and don't place it to rawFront
                     *   T = C  - it is recursive topology(recursion in same layer)
                     *   T = C, but t=c  - it is recurrent link over 1 neuron
                     *   T = C + 1 - T already handled. We should add T to rawFront
                     *   T > C + 1 - latency to T is > 1. Don't touch T.
                     */

                    //Assume that when cache becomes coherent, all neuron[hops] vars become zero
                    if(hops[toNeuron] == 0)
                        hops[toNeuron] = hops[currentNeuron] + link.getLatency();

                    if(hops[toNeuron] == hops[currentNeuron] + 1)
                        rawFront.push_back(toNeuron); 

                } //BOOST_FOREACH( Link )

            } //rawFront iteration ( Neuron )

            //new rawFront formed
            formatFront(rawFront);

        } while( rawFront.size() > 0 );

        //We rebuilded cache
        cache.fixed();

    } //regenerateCache

    void
    Net::threadBase(Runner* _runner, const Net* _net, unsigned _cur_thread, 
                                                      boost::barrier* _barrier)
    {
        RunDirection dir = _runner->getDirection();
        const NetCache& _cache = _net->getCache();

        unsigned layer;
        (dir == ForwardRun) ?  (layer = 0) : (layer = _cache.layers.size() - 1);

        const unsigned threads = _net->getWorkThreadsCount();

        do {
            //Process current layer
            for(unsigned i = _cur_thread; i < _cache.layers[layer].size(); i += threads)
                _runner->run( *_cache.layers[layer][i], *_net ); //We pass Net* to runner, because 
                                                                //learning algorithms require 
                                                                //read-only access to Net attributes

            std::cout<<_cur_thread<<std::endl;
            //Wait for other threads
            _barrier->wait();
        } while( (dir == ForwardRun && ++layer < _cache.layers.size()) || (dir == BackwardRun && layer-- > 0) );
        /*
         * A little comment.
         * Cache structure:
         * Layer1:   thread1_data, thread2_data, ...
         * Layer2:   thread1_data, thread2_data, ...
         * ...
         * LayerN-1: thread1_data, thread2_data, ...
         * LayerN:          <= last layer. ALWAYS empty! (see Net::regenerateCache())
         * cache.size() == N+1;
         * thread_data is vector of Neuron*
         */
    } //threadBase

}; //pann

File: trunk/pypann/pypann/Core/Net.py
#
# NetCache
#
class NetCache:
    def __init__(self):
        self._ok = False
        self.layers = []

    def is_ok(self):
        return self._ok

    def invalidate(self):
        self._ok = False

    def fix(self):
        self._ok = True
#
# Barrier
#
from threading import Condition

class Barrier:
    def __init__(self, threads):
        self.threads = self.remains = threads
        self.cond = Condition()

    def wait(self):
        with self.cond:
            self.remains -= 1
            if self.remains:
                self.cond.wait()
            else:
                self.remains = self.threads
                self.cond.notifyAll()

#
# Net
#
from multiprocessing import cpu_count
from threading import Thread
from Link import Link
from Weight import Weight

class Net:
    def __init__(self):
        self.input_neurons = []
        self.worker_threads_count = 1 #cpu_count()
        self._cache = NetCache()

    def add_input_neuron(self, n):
        self.input_neurons.append(n)
        self._cache.invalidate()

    def remove_neuron(self, n):
        for link in n._links_in:  self.disconnect(link.to(), n)
        for link in n._links_out: self.disconnect(n, link.to())

        try:
            self.input_neurons.remove(n)
        except ValueError:
            pass

        del n
        self._cache.invalidate()

    def connect(self, n_from, n_to, w = None, latency = 1):
        if w == None:
            w = Weight()
        n_from._links_out.append(Link(n_to, w, latency))
        n_to._links_in.append(Link(n_from, w, latency))
        self._cache.invalidate()

    def disconnect(self, n_from, n_to):
        n_from._links_out = [link for link in n_from._links_out if link.to() != n_to]
        n_to._links_in = [link for link in n_to._links_in if link.to() != n_from]
        self._cache.invalidate()

    def set_input(self, values):
        for i in range(len(self.input_neurons)):
            self.input_neurons[i].input = values[i]

    def get_output(self):
        if not self._cache.is_ok():
            self._update_cache()

        if len(self._cache.layers):
            return map(lambda n: n.output, self._cache.layers[-1])
        else:
            return []

    def run(self, runner = lambda x, y: x.run(), dir = True):
        """
        Use runner on all neurons

        runner := func(neuron, net)
        dir := True | False
        if dir == True then perform feedforward run, else - backward run
        """
        def worker(id):
            for layer in self._cache.layers:
                for i in range(id, len(layer), self.worker_threads_count):
                    runner(layer[i], self)
                barrier.wait()

        if not self._cache.is_ok():
            self._update_cache()

        if not dir:
            self._cache.layers.reverse()

        thread_pool = []
        barrier = Barrier(self.worker_threads_count)
        for i in range(self.worker_threads_count):
            t = Thread(target = worker, args = [i])
            t.daemon = True
            t.start()
            thread_pool.append(t)

        for t in thread_pool:
            t.join()

        if not dir:
            self._cache.layers.reverse()

    def _update_cache(self):
        """
        Rebuild internal cache on demand
        """
        self._cache.layers = []
        hops = dict(map(lambda x: (x, 1), self.input_neurons))

        layer = set(self.input_neurons)
        while len(layer):
            self._cache.layers.append(list(layer))
            next_layer = set()
            for n in layer:
                for link in n._links_out:
                    if not link.to() in hops:
                        hops[link.to()] = hops[n] + link.latency()
                    if hops[link.to()] == hops[n] + 1:
                        next_layer.add(link.to())
                layer = next_layer

        self._cache.fix()

